File: etl/test/report_extractor_test.rb
require "minitest/autorun"
require_relative "../lib/etl/report_extractor"

class ReportExtractorTest < Minitest::Test
  SAMPLE_PATH = "ignored.docx"
  SAMPLE_PARAGRAPHS = [
    "Complete Wellbeing Health Assessment",
    "Gender: male (46)",
    "Personal Report",
    "Client reports stress in daily life.",
    "Client reports poor sleep quality.",
    "Summary of Key Results",
    "Nutrition section of the questionnaire: Mediterranean Style of Diet"
  ].freeze

  SAMPLE_TABLE = [
    ["Measurement", "Value"],
    ["Height (cm)", "x"],
    ["Blood Pressure (mmHg)", "x / x"]
  ].freeze

  def with_stubbed_docx(paragraphs: SAMPLE_PARAGRAPHS, table: SAMPLE_TABLE)
    fake_doc = Struct.new(:paragraphs, :tables).new(paragraphs, [table])

    Etl::DocxDocument.stub(:new, fake_doc) do
      yield
    end
  end

  def test_extracts_gender_and_age_from_the_sample_report
    with_stubbed_docx do
      report = Etl::ReportExtractor.extract(SAMPLE_PATH)

      assert_equal "Male", report.gender
      assert_equal 46, report.age
    end
  end

  def test_extracts_narrative_sections_for_tagging
    with_stubbed_docx do
      report = Etl::ReportExtractor.extract(SAMPLE_PATH)

      assert_includes report.personal_report_text, "stress"
      assert_includes report.personal_report_text, "sleep"
      assert_includes report.nutrition_text, "Mediterranean Style of Diet"
    end
  end

  def test_summary_table_labels_are_found_even_when_unfilled
    with_stubbed_docx do
      report = Etl::ReportExtractor.extract(SAMPLE_PATH)

      # The sample report is a template with placeholder ("x") values, so the
      # numeric fields correctly resolve to nil rather than raising.
      assert_nil report.height_cm
      assert_nil report.blood_pressure_systolic
    end
  end
end

File: desktop/etl/test/reportExtractor.test.js
"use strict";

const { test, before, after } = require("node:test");
const assert = require("node:assert/strict");
const fs = require("node:fs");
const os = require("node:os");
const path = require("node:path");

const { extractReport } = require("../reportExtractor");
const { buildSampleDocxBuffer } = require("./fixtures/buildSampleDocx");

let tmpDir;
let samplePath;

before(async () => {
  tmpDir = fs.mkdtempSync(path.join(os.tmpdir(), "wellbeing-docx-"));
  samplePath = path.join(tmpDir, "Complete Wellbeing Sample 01012023.docx");
  fs.writeFileSync(samplePath, await buildSampleDocxBuffer());
});

after(() => {
  fs.rmSync(tmpDir, { recursive: true, force: true });
});

// Same assertions as ../../../etl/test/report_extractor_test.rb, against an
// equivalent hand-built fixture (the real sample docx is intentionally not
// present in this environment).
test("extracts gender and age from the sample report", async () => {
  const report = await extractReport(samplePath);
  assert.equal(report.gender, "Male");
  assert.equal(report.age, 46);
});

test("extracts narrative sections for tagging", async () => {
  const report = await extractReport(samplePath);
  assert.match(report.personal_report_text, /stress/);
  assert.match(report.personal_report_text, /sleep/);
  assert.match(report.nutrition_text, /Mediterranean Style of Diet/);
});

test("summary table labels are found even when unfilled", async () => {
  const report = await extractReport(samplePath);
  // "Height (cm)" has the placeholder value "x" in the fixture (as in the
  // real template), so it should resolve to null rather than throwing/NaN.
  assert.equal(report.height_cm, null);
  assert.equal(report.blood_pressure_systolic, 120);
  assert.equal(report.blood_pressure_diastolic, 80);
  assert.equal(report.hdl_cholesterol, 1.5);
});
